server/src/main/java/org/opensearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -388,34 +388,57 @@ protected BytesRef indexedValueForSearch(Object value) {
         }
 
         @Override
-        public Query termsQuery(List<?> values, QueryShardContext context) {
+        public Query termsQuery(List<?> values, RewriteOverride rewriteOverride, QueryShardContext context) {
```

Comment:
The `RewriteOverride` definitely belongs to `QueryRewriteContext`  (which `QueryShardContext` implements)

Reply:
Sorry, should have been more specific, I think we could also move the argument ` RewriteOverride rewriteOverride` to `QueryRewriteContext` (as a field), so the change itself would be non-breaking (so signature changes for public methods), thank you

Reply:
Ah, I understand. But how will users be able to explicitly set the value? I don't believe we have access to `QueryShardContext` at search time?

Reply:
> Ah, I understand. But how will users be able to explicitly set the value? 
I think the value comes from query (or to be precise, `QueryBuilder`), right?  I was thinking when `DefaultSearchContext` constructs `QueryShardContext`, it has access to `request.source().query()` that would have the override property available. Do you think it makes sense? thank you

Reply:
For clarity, let's say we have the following request:

```
query:
  terms: {
    some_keyword: ["400", "5"],
    rewrite_override: "doc_values_only"
  }

```

This is how we're able to specify the override parameter. `request.source().query()` points to an instance of `QueryBuilder`. So for us to consume this new `rewrite_override` parameter we would need to pass it into any query builder(here we do it in all MultiTerm query builder - terms, range, etc). That is the only way we have access to the parameter, I believe `QueryShardContext` contains only logic on how to execute the query on the shard and not on how to read the parameters of the query. let me know if this makes sense, thanks!

Reply:
> So for us to consume this new rewrite_override parameter we would need to pass it into any query builder(here we do it in all MultiTerm query builder - terms, range, etc). T

That could be done with interface, fe `SupportsQueryRewriteOverride` so only some `QueryBuilder`s would implement it.

> I believe QueryShardContext contains only logic on how to execute the query on the shard and not on how to read the parameters of the query. 

Correct, it is just carrier of the `rewrite_override` that could be passed as constructor argument. 

The primary reason to use `QueryShardContext` to hold `rewrite_override` is to keep APIs (public methods) unchanged, otherwise we will won't be able to backport it to `2.x`. Also, `rewrite_override` looks concise (to me0 with `QueryRewriteContext` (that `QueryShardContext` inherits)

Reply:
Hmm @reta I thought about this a little more and I feel like we would want the rewrite_override method to target only certain APIs, terms, range etc and only for the keyword field. Hence, it's cleaner to only do this at the query builder level. Touching the shard context would mean this would spread to other queries.

Plus on you point of backporting, we're only adding new parameters to the query and not changing any existing logic so backporting should be fine. wdyt?

Reply:
> Hence, it's cleaner to only do this at the query builder level. Touching the shard context would mean this would spread to other queries.

Thanks @harshavamsi , this is a good point, let's keep it as overloaded method, as you did, thank you